File: src/main/java/refactown/cleancode/c01cleancode/Refactoring1.java
package refactown.cleancode.c01cleancode;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

import static java.lang.reflect.Modifier.isFinal;
import static java.lang.reflect.Modifier.isStatic;

public class Refactoring1 {

    public static int countIntConstants(Class<?> clazz) {
        if (clazz == null || clazz.isEnum() || Modifier.isAbstract(clazz.getModifiers())) {
            return -1;
        }
        List<Field> fields = new ArrayList<>();
        for (Field field : clazz.getDeclaredFields()) {
            fields.add(field);
        }
        Class<?> superClass = clazz.getSuperclass();
        while (superClass != null && !superClass.equals(Object.class)) {
            for (Field field : superClass.getDeclaredFields()) {
                fields.add(field);
            }
            superClass = superClass.getSuperclass();
        }
        int count = 0;
        Class<?> fieldType;
        for (Field field : fields) {
            fieldType = field.getType();
            if (fieldType.equals(Integer.class) || fieldType.equals(int.class)
                    || fieldType.equals(Long.class)	|| fieldType.equals(long.class)
                    || fieldType.equals(Short.class) || fieldType.equals(short.class)
                    || fieldType.equals(Byte.class) || fieldType.equals(byte.class)
                    || fieldType.equals(BigInteger.class)) {
                if (isStatic(field.getModifiers()) && isFinal(field.getModifiers())) {
                    System.out.println("Int Constant:" + field);
                    count++;
                }
            }
        }
        return count;
    }
}

File: src/main/kotlin/refactown/cleancode/c01cleancode/Refactoring1.kt
package refactown.cleancode.c01cleancode

import java.lang.reflect.Field
import java.lang.reflect.Modifier
import java.math.BigInteger
import java.util.*

fun countIntConstants(clazz: Class<*>?): Int {
    if (clazz != null && !clazz.isEnum && !Modifier.isAbstract(clazz.modifiers)) {
        val fields = ArrayList<Field>()
        for (field in clazz.declaredFields) {
            fields.add(field)
        }
        var superclass: Class<*>? = clazz.superclass
        while (superclass != null && superclass != Any::class.java) {
            for (field in superclass.declaredFields) {
                fields.add(field)
            }
            superclass = superclass.superclass
        }
        var count = 0
        var type: Class<*>
        for (field in fields) {
            type = field.getType()
            if (type == Int::class.java || type == Int::class.javaPrimitiveType
                    || type == Long::class.java || type == Long::class.javaPrimitiveType
                    || type == Short::class.java || type == Short::class.javaPrimitiveType
                    || type == Byte::class.java || type == Byte::class.javaPrimitiveType
                    || type == BigInteger::class.java) {
                if (Modifier.isStatic(field.getModifiers())
                        && Modifier.isFinal(field.getModifiers())) {
                    println("Int Constant:$field"); count++
                }
            }
        }
        return count
    } else return -1
}
